orttraining/orttraining/core/graph/graph_parser/pattern_graph.h

```diff
@@ -0,0 +1,299 @@
+    if (!g && !p)
+      return true;
+    if (!g && p || !p && g)
+      return false;
+    if (!skip_optype_ && p->NameEquals(g->OpType())) {
+      // PARSER_LOG << "OpType mismatch, "
+      //            << "g is: " << g->OpType() << ", p is: " << p->OpType();
+      return false;
+    }
+    if (!skip_domain_ && p->DomainsContain(g->Domain())) {
+      // PARSER_LOG << "Domain mismatch, "
+      //            << "g is: " << g->Domain() << ", p is: " << p->Domain();
+      return false;
+    }
+    if (!skip_version_ && p->VersionsContain(g->SinceVersion())) {
+      // PARSER_LOG << "Version mismatch, "
+      //            << "g is: " << g->SinceVersion() << ", p is: " << p->SinceVersion();
+      return false;
+    }
+    return true;
+  }
+
+  bool skip_optype_, skip_domain_, skip_version_, skip_path_;
+};
+
+// typedef std::function<bool(const Node*, const PNode*, const Graph&, const PatternGraph&)> NodeCompareFunc;
+
+struct PatternGraph {
+  PatternGraph(const std::vector<NodeDef>& constants, const std::vector<PNode>& pnodes,
+               const std::string& root_node = "", const std::string& name = "")
+      : name_(name),
+        root_node_(root_node),
+        nodes(constants) {
+    ort_model_ptr_ = std::make_unique<Model>("pattern_model", false, logging::LoggingManager::DefaultLogger());
+    for (auto pnode : pnodes) {
+      name_pnode_mapping_[pnode.GetNodeName()] = &pnode;
+      nodes.push_back(pnode.GetNodeDef());
+      domain_map[pnode.GetNodeName()] = pnode.GetValidDomains();
+      version_map[pnode.GetNodeName()] = pnode.GetValidVersions();
+    }
+
+    ORT_ENFORCE(ToGraphInternal().IsOK());
+
+    default_compare_func_ = std::make_unique<DefaultNodeCompareFunc>(false, false, false, false);
+  }
+
+  const std::string& name() const {
+    return name_;
+  }
+
+  // TODO: should hide this in private field.
+  const std::string& root_node() const { return root_node_; }
+
+  PatternGraph& SetCustomConstraint(std::unique_ptr<NodeCompareFunc> func, std::string node_name = "") {
+    if (node_name.empty()) {
+      default_compare_func_ = std::move(func);
+      custom_constraints_.clear();
+    } else {
+      custom_constraints_[node_name] = std::move(func);
+    }
+    return *this;
+  }
+
+  // TODO: evaluate should we have them?
+  /*
+  PatternGraph& add_node(NodeDef node) {
+    nodes.push_back(node);
+    return *this;
+  }
+
+  PatternGraph& add_nodes(const std::vector<NodeDef>& nodes) {
+    this->nodes.insert(this->nodes.end(), nodes.begin(), nodes.end());
+    return *this;
+  }
+
+
+  PatternGraph& disable_optype_check() {
+    check_optype = false;
+    return *this;
+  }
+
+  PatternGraph& disable_domain_check() {
+    check_domain = false;
+    return *this;
+  }
+
+  PatternGraph& disable_version_check() {
+    check_version = false;
+    return *this;
+  }
+
+  PatternGraph& disable_path_check() {
+    check_path = false;
+    return *this;
+  }
+  */
```

Comment:
I am thinking we should avoid complexity for user to use the APIs. there are a few here, and there will be more. So I refactored to, using a single entry "SetCustomConstraint" to customize node filtering.